File: apps/desktop/src/main/observe/sessionStatus.ts
import {
  type Agent,
  derived,
  isSteerable,
  openTurn,
  rootAgent,
  type SessionPosture,
  type SessionStatus,
  type StopReason,
  type Tiered,
} from '../../shared'
import { ASK_TOOL } from './toolCalls'

// A live process can sit quiet mid-tool; only a transcript touched within this window
// corroborates an actively-running session — beyond it, ambiguity resolves down.
export const RECENT_ACTIVITY_MS = 10 * 60 * 1000

// The stop reasons that mean the agent HALTED rather than finished. `end_turn` is a completed
// exchange; these three are a wall it hit, which is a different thing to tell you about.
const HALTING_REASONS: readonly StopReason[] = ['max_tokens', 'max_turn_requests', 'refusal']

export interface StatusSignals {
  posture: SessionPosture
  /** A live CLI process was matched on this Session's cwd. */
  processMatch: boolean
  lastTimestampMs: number | null
  nowMs: number
  /** The runtime tree, empty when the transcript could not be parsed into one. */
  agents: Agent[]
}

function isRecent({ lastTimestampMs, nowMs }: StatusSignals): boolean {
  return lastTimestampMs !== null && nowMs - lastTimestampMs <= RECENT_ACTIVITY_MS
}

/**
 * A pending `AskUserQuestion` in the Turn still running is the ONLY honest `asking` for a session
 * Argo does not own: the tool call landed in the transcript and no result answered it. A resolved
 * question, or one sitting in a Turn that has since ended, reads as `idle` — a false `asking` is
 * a false-active, so ambiguity degrades down. An agent's free-form question is indistinguishable
 * from idle in the record and is never promoted here.
 */
function hasPendingAsk(root: Agent): boolean {
  const turn = openTurn(root)
  if (turn === null) return false
  return turn.toolCalls.some((call) => call.name === ASK_TOOL && call.status === 'pending')
}

// `ended` needs a process exit Argo WITNESSED, and the one it witnesses is its own PTY dying —
// which is precisely the `orphaned` posture (claimed, PTY gone). Everything else floors at
// `idle`: a session Argo never owned may simply be sitting quiet, and calling that a shutdown
// would be a termination nobody observed.
function quietStatus(signals: StatusSignals, live: boolean): SessionStatus {
  return !live && signals.posture === 'orphaned' ? 'ended' : 'idle'
}

function statusOf(signals: StatusSignals, live: boolean): SessionStatus {
  const root = rootAgent(signals.agents)
  // No Turns to read (an unparseable body, or a session that has only just started) leaves
  // liveness UNTOUCHED: the process match plus recency is the reading, and only the refinements
  // a segmented tree would add go absent (cockpit-failure-states-spec.md §8).
  if (root === null || root.turns.length === 0) {
    return live ? 'running' : quietStatus(signals, live)
  }
  if (live) {
    if (hasPendingAsk(root)) return 'asking'
    if (openTurn(root) !== null) return 'running'
  }
  // `stopped` is managed-only: reading a wall the agent hit as a wall (rather than as a quiet
  // idle) is a claim about a session Argo owns, and CONTEXT.md L2 collapses it to `idle` the
  // moment the same session is observed from the outside.
  const halted = root.turns.at(-1)?.stopReason ?? null
  if (isSteerable(signals.posture) && halted !== null && HALTING_REASONS.includes(halted)) {
    return 'stopped'
  }
  return quietStatus(signals, live)
}

/**
 * The Session's status, honesty-gated by posture (CONTEXT.md L2). External and orphaned floor at
 * `running · asking? · idle · ended`; only a managed Session can reach `stopped`, and `permission`
 * awaits the companion-plugin channel that carries a prompt no transcript holds.
 *
 * The tier is DERIVED for every posture, managed included. Argo owns the PTY, but it does not own
 * the link from that PTY to a transcript file — the match is cwd plus a time window, which is not
 * a unique key. Grading this DIRECT off ownership alone would be exactly the false DIRECT the
 * degrade-down rule exists to prevent.
 */
export function deriveSessionStatus(signals: StatusSignals): Tiered<SessionStatus> {
  return derived(statusOf(signals, signals.processMatch && isRecent(signals)))
}

File: apps/macOS/Packages/ArgoEngine/Sources/ArgoEngine/Session/SessionStatus.swift
/// What a Session is doing, rolled up (CONTEXT.md, "Session status"). Always DERIVED, never
/// stored, and carrying its own `unknown` so a status that cannot be established honestly has
/// somewhere to go (ADR-0008's degrade-down rule).
public enum SessionStatus: Sendable, Equatable, CaseIterable {
    /// Argo started the process and it has not spoken yet. DIRECT, managed only, and unreachable
    /// from `read(_:)` — no record exists to read it off (`HubSession.statusReading`, #587).
    case starting
    /// A Turn is in progress.
    case running
    /// Blocked on an agent `request_permission` prompt. DIRECT, managed only.
    case permission
    /// Blocked on a structured `AskUserQuestion`.
    case asking
    /// A Turn ended `end_turn`, or there is no live signal. Includes an agent's free-form question,
    /// which is indistinguishable from idle in the record.
    case idle
    /// A Turn ended `max_tokens`, `max_turn_requests` or `refusal`.
    case stopped
    /// The Session terminated.
    case ended
    /// Nothing establishes a status. Not a state the Session is in — a state we cannot claim.
    case unknown
}

public extension SessionStatus {
    /// The Session's status, honesty-gated by posture and corroborated by liveness.
    ///
    /// DERIVED for every posture, managed included: Argo owns the PTY but not the link from it to a
    /// transcript file, which is a working directory plus a time window and not a unique key. A
    /// Turn Argo itself submitted needs no such link and is read above this, at DIRECT
    /// (`HubSession.statusReading`, #1048) — every reading that reaches HERE is DERIVED.
    ///
    /// `permission` is unreachable here by construction — it belongs to the channels Argo owns (the
    /// permission gate, then the companion's report).
    static func read(_ signals: SessionSignals) -> SessionStatusReading {
        SessionStatusReading(tier: .derived, status: status(signals))
    }

    /// Liveness is what may claim a Session is working; the record is what says what it was doing.
    ///
    /// A Session with no Turn in it yet leaves liveness UNTOUCHED: the process match plus recency
    /// IS the reading, and only the refinements a parsed record would add go absent.
    private static func status(_ signals: SessionSignals) -> SessionStatus {
        guard signals.liveness == .live else { return quiet(signals) }
        if signals.pendingAsk, signals.turnOpen {
            return .asking
        }
        return signals.turnOpen || !signals.hasTurns ? .running : boundary(signals)
    }

    /// A Session nothing corroborates as working. `ended` needs a process exit Argo WITNESSED,
    /// which is exactly the `orphaned` posture; everything else falls back to the last boundary.
    private static func quiet(_ signals: SessionSignals) -> SessionStatus {
        signals.provenance == .orphaned ? .ended : boundary(signals)
    }

    /// What the last Turn boundary says on its own.
    ///
    /// `stopped` is managed-only: reading a wall the agent hit AS a wall is a claim about a Session
    /// Argo owns, and it collapses to `idle` the moment the same Session is observed from outside.
    private static func boundary(_ signals: SessionSignals) -> SessionStatus {
        // A Turn opened since the last boundary, so that boundary is the previous Turn's and says
        // nothing about this one. Uncorroborated, an open Turn is never `running` — and never
        // `idle` either: the record says a Turn was started and nothing says it finished, so the
        // calm word would be the one a FINISHED Session draws and a wrong liveness would be
        // invisible on the row. `unknown` is the honest outline for it (ADR-0008, #1261).
        guard !signals.turnOpen else { return .unknown }
        return switch signals.lastStop {
        case .endTurn, .cancelled: .idle
        case .maxTokens, .maxTurnRequests, .refusal:
            signals.provenance == .managed ? .stopped : .idle
        case .unknown, .none: .unknown
        }
    }
}

public extension SessionStatus {
    /// Whether the CLI's own prompt is free to take a line somebody types at it (#1217).
    ///
    /// `running` is the obvious one: the CLI queues a line typed mid-Turn as the NEXT prompt rather
    /// than running it. The other two matter more. A `permission` or an `asking` Session has its
    /// keyboard held by a DIALOG, so a line typed then is eaten by that dialog — and the Return
    /// behind it answers whatever the dialog had highlighted. That is worse than a line going
    /// nowhere, which is why the three answer together.
    ///
    /// `starting` is deliberately NOT among them. Argo has written the argv and has not heard the
    /// child yet; the prompt is on its way rather than held by something else, and refusing there
    /// would refuse the one moment a Session is being set up in.
    ///
    /// This is the question about a PROMPT. A slash command is a different question and has its
    /// own answer below — see `takesSlashCommand` for why one reading could not serve both.
    var takesTypedLine: Bool {
        switch self {
        case .running, .permission, .asking: false
        case .starting, .idle, .stopped, .ended, .unknown: true
        }
    }

    /// Whether a SLASH COMMAND typed at this Session would be run rather than swallowed (#1658).
    ///
    /// Narrower than `takesTypedLine` by exactly one status, and the difference is the whole point.
    /// `/rename`, `/model` and `/effort` are handled by the harness itself and never reach the
    /// agent, so a Turn in flight does not queue them the way it queues a prompt: measured on a
    /// real PTY, `/rename` typed mid-Turn renamed the Session while the Turn ran on untouched.
    /// Reading `running` as a refusal cost every Ticket-started Session its name for the WHOLE
    /// length of its run, which is the only stretch anybody is looking at the list.
    ///
    /// A DIALOG is still a refusal, and for the reason it always was: under a `permission` or an
    /// `asking` the keyboard belongs to that dialog, so the line is eaten by it and the Return
    /// behind the line answers whatever it had highlighted (#1217). That risk is about who owns
    /// the keyboard, not about what the line says, so it survives the split unchanged.
    var takesSlashCommand: Bool {
        switch self {
        case .permission, .asking: false
        case .running, .starting, .idle, .stopped, .ended, .unknown: true
        }
    }
}
